File: internal/trace/trace.go
// Package trace ports validators/trace.py (stdlib only): the
// Requirement -> Decision -> Implementation -> Test -> Verification
// Result index plus the trace CLI surface data.
package trace

import (
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"unicode/utf8"

	"github.com/shiploom/ai-builder/internal/validate"
)

// Relations in fixed order (mirrors RELATIONS).
var Relations = []string{"requires", "decided_by", "implemented_by", "tested_by", "verified_by"}

// EmptyLinks mirrors empty_links().
func EmptyLinks() map[string][]string {
	return map[string][]string{
		"requires": {}, "decided_by": {}, "implemented_by": {},
		"tested_by": {}, "verified_by": {},
	}
}

// Trace maps artifact id -> relation -> targets.
type Trace map[string]map[string][]string

// BuildTrace mirrors build_trace(). Returns (trace, errors, warnings);
// errors/warnings are unsorted (callers sort, like trace.main).
func BuildTrace(schemas *validate.Schemas, target string, strict bool) (Trace, []validate.Entry, []validate.Entry) {
	var errors []validate.Entry
	var warnings []validate.Entry
	trace := Trace{}
	root := target
	base := target
	if info, err := os.Stat(target); err == nil && !info.IsDir() {
		base = filepath.Dir(target)
	}
	files := validate.CollectFiles(root)
	for _, path := range files {
		if validate.IsExcluded(path, base) {
			continue
		}
		if filepath.Ext(path) == ".md" {
			raw, err := os.ReadFile(path)
			if err != nil {
				errors = append(errors, validate.Entry{Path: path,
					Message: fmt.Sprintf("unreadable: %s", fsErrText(err))})
				continue
			}
			text, ok := utf8String(raw)
			if !ok {
				errors = append(errors, validate.Entry{Path: path,
					Message: fmt.Sprintf("unreadable: invalid UTF-8 in %s", path)})
				continue
			}
			fm, _, perr := validate.ParseFrontmatter(text)
			if fm == nil {
				if perr != "" {
					errors = append(errors, validate.Entry{Path: path,
						Message: perr, Rule: "frontmatter"})
				}
				continue
			}
			if _, hasID := fm["id"]; !hasID {
				continue
			}
			if _, hasKind := fm["kind"]; !hasKind {
				continue
			}
			aid, ok := fm["id"].(string)
			if !ok {
				continue
			}
			if _, dup := trace[aid]; dup {
				errors = append(errors, validate.Entry{Path: path,
					Message: fmt.Sprintf("duplicate artifact id %s", validate.PyRepr(aid)),
					Rule:    "trace.duplicate"})
				continue
			}
			links := EmptyLinks()
			if rawLinks, ok := fm["links"].(map[string]any); ok {
				for _, rel := range Relations {
					if vals, ok := rawLinks[rel].([]any); ok {
						for _, v := range vals {
							if s, ok := v.(string); ok {
								links[rel] = append(links[rel], s)
							}
						}
					}
					if links[rel] == nil {
						links[rel] = []string{}
					}
				}
			}
			trace[aid] = links
		} else if filepath.Ext(path) == ".json" && filepath.Base(path) != "trace.json" {
			raw, err := os.ReadFile(path)
			if err != nil {
				continue
			}
			doc, err := validate.DecodeJSON(raw)
			if err != nil {
				continue
			}
			if obj, ok := doc.(map[string]any); ok {
				if _, hasSchema := obj["$schema"]; hasSchema {
					continue
				}
			}
			var objs []any
			if list, ok := doc.([]any); ok {
				objs = list
			} else {
				objs = []any{doc}
			}
			for _, item := range objs {
				obj, ok := item.(map[string]any)
				if !ok {
					continue
				}
				if _, hasR := obj["results"]; hasR {
					if _, hasV := obj["verdict"]; hasV {
						if rid, ok := obj["id"].(string); ok {
							if _, exists := trace[rid]; !exists {
								var tested []string
								if results, ok := obj["results"].([]any); ok {
									for _, r := range results {
										if robj, ok := r.(map[string]any); ok {
											if acc, ok := robj["acceptanceId"].(string); ok {
												tested = append(tested, acc)
											}
										}
									}
								}
								if tested == nil {
									tested = []string{}
								}
								entry := EmptyLinks()
								entry["tested_by"] = tested
								trace[rid] = entry
							}
						}
					}
				} else if _, hasH := obj["howToVerify"]; hasH {
					if id, ok := obj["id"].(string); ok {
						if _, exists := trace[id]; !exists {
							trace[id] = EmptyLinks()
						}
					}
				}
			}
		}
	}
	schema, err := schemas.Load("trace")
	if err == nil {
		for _, msg := range validate.ValidateAgainstSchema(traceToAny(trace), schema, "$") {
			errors = append(errors, validate.Entry{Path: root, Message: msg, Rule: "trace.schema"})
		}
	}
	known := map[string]bool{}
	for aid := range trace {
		known[aid] = true
	}
	aids := make([]string, 0, len(trace))
	for aid := range trace {
		aids = append(aids, aid)
	}
	sort.Strings(aids)
	for _, aid := range aids {
		for _, rel := range Relations {
			for _, tgt := range trace[aid][rel] {
				if !known[tgt] {
					msg := fmt.Sprintf("%s: dangling link %s -> %s",
						aid, rel, validate.PyRepr(tgt))
					entry := validate.Entry{Path: root, Message: msg, Rule: "links.dangling"}
					if strict {
						errors = append(errors, entry)
					} else {
						warnings = append(warnings, entry)
					}
				}
			}
		}
	}
	return trace, errors, warnings
}

func traceToAny(trace Trace) map[string]any {
	out := map[string]any{}
	for aid, links := range trace {
		m := map[string]any{}
		for rel, targets := range links {
			list := make([]any, 0, len(targets))
			for _, t := range targets {
				list = append(list, t)
			}
			m[rel] = list
		}
		out[aid] = m
	}
	return out
}

func utf8String(raw []byte) (string, bool) {
	if !utf8.Valid(raw) {
		return "", false
	}
	return string(raw), true
}

func fsErrText(err error) string {
	if pe, ok := err.(*os.PathError); ok {
		return pe.Err.Error()
	}
	return err.Error()
}

File: validators/trace.py
#!/usr/bin/env python3
"""Shiploom trace index builder (stdlib-only).

Builds the Requirement -> Decision -> Implementation -> Test ->
Verification Result index from artifact frontmatter links plus acceptance
and verification-report JSON ids (MASTER_SPEC 16, BUILD_PLAN PR3).

Usage:
    python3 validators/trace.py [--strict] [path] [--out trace.json]

Output (stdout, always JSON):
    {"ok": bool, "trace": {...}, "errors": [...], "warnings": [...]}

--out writes the trace map to FILE (parents created). Exit 0 pass, 2 fail.
No network. No other writes.
"""

import argparse
import json
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent.parent))

from validators.validate import (  # noqa: E402
    _is_excluded,
    collect_files,
    load_schema,
    parse_frontmatter,
    validate_against_schema,
)

RELATIONS = ("requires", "decided_by", "implemented_by", "tested_by", "verified_by")


def empty_links():
    return {rel: [] for rel in RELATIONS}


def build_trace(target, strict=False):
    """Build and schema-check the trace index.

    Returns (trace, errors, warnings). `trace` maps artifact id ->
    {requires, decided_by, implemented_by, tested_by, verified_by}.
    """
    errors, warnings, trace = [], [], {}
    root = Path(target)
    base = root if root.is_dir() else root.parent
    files = collect_files(root)

    for path in files:
        if _is_excluded(path, base):
            continue
        if path.suffix == ".md":
            try:
                text = path.read_text(encoding="utf-8")
            except OSError as exc:
                errors.append({"path": str(path), "message": "unreadable: %s" % exc})
                continue
            fm, _, perr = parse_frontmatter(text)
            if fm is None:
                if perr:
                    errors.append({"path": str(path), "message": perr, "rule": "frontmatter"})
                continue
            if "id" not in fm or "kind" not in fm:
                continue
            aid = fm.get("id")
            if not isinstance(aid, str):
                continue
            if aid in trace:
                errors.append({"path": str(path),
                               "message": "duplicate artifact id %r" % aid,
                               "rule": "trace.duplicate"})
                continue
            links = empty_links()
            raw_links = fm.get("links") or {}
            if isinstance(raw_links, dict):
                for rel in RELATIONS:
                    vals = raw_links.get(rel, [])
                    links[rel] = list(vals) if isinstance(vals, list) else []
            trace[aid] = links
        elif path.suffix == ".json" and path.name != "trace.json":
            try:
                doc = json.loads(path.read_text(encoding="utf-8"))
            except (OSError, ValueError):
                continue  # JSON syntax is validate.py's job
            if isinstance(doc, dict) and "$schema" in doc:
                continue
            objs = doc if isinstance(doc, list) else [doc]
            for obj in objs:
                if not isinstance(obj, dict):
                    continue
                if "results" in obj and "verdict" in obj and isinstance(obj.get("id"), str):
                    rid = obj["id"]
                    if rid not in trace:
                        tested = [r.get("acceptanceId") for r in obj.get("results", [])
                                  if isinstance(r, dict) and isinstance(r.get("acceptanceId"), str)]
                        trace[rid] = dict(empty_links(), tested_by=tested)
                elif "howToVerify" in obj and isinstance(obj.get("id"), str):
                    if obj["id"] not in trace:
                        trace[obj["id"]] = empty_links()

    for msg in validate_against_schema(trace, load_schema("trace"), "$"):
        errors.append({"path": str(root), "message": msg, "rule": "trace.schema"})

    known = set(trace)
    for aid in sorted(trace):
        for rel in RELATIONS:
            for tgt in trace[aid][rel]:
                if tgt not in known:
                    msg = "%s: dangling link %s -> %r" % (aid, rel, tgt)
                    entry = {"path": str(root), "message": msg, "rule": "links.dangling"}
                    (errors if strict else warnings).append(entry)

    return trace, errors, warnings


def main(argv=None):
    parser = argparse.ArgumentParser(description="Shiploom trace index builder (stdlib-only).")
    parser.add_argument("path", nargs="?", default=".", help="file or directory (default: .)")
    parser.add_argument("--strict", action="store_true",
                        help="dangling links fail instead of warn")
    parser.add_argument("--out", default=None, help="write trace map to FILE")
    args = parser.parse_args(argv)

    if not Path(args.path).exists():
        result = {"ok": False, "trace": {},
                  "errors": [{"path": args.path, "message": "no such file or directory"}],
                  "warnings": []}
        json.dump(result, sys.stdout, indent=2, sort_keys=True)
        sys.stdout.write("\n")
        return 2

    trace, errors, warnings = build_trace(args.path, strict=args.strict)
    errors.sort(key=lambda e: (e["path"], e["message"]))
    warnings.sort(key=lambda e: (e.get("path", ""), e.get("message", "")))
    result = {"ok": not errors, "trace": trace, "errors": errors, "warnings": warnings}
    if args.out:
        out = Path(args.out)
        out.parent.mkdir(parents=True, exist_ok=True)
        out.write_text(json.dumps(trace, indent=2, sort_keys=True) + "\n", encoding="utf-8")
    json.dump(result, sys.stdout, indent=2, sort_keys=True)
    sys.stdout.write("\n")
    return 0 if not errors else 2


if __name__ == "__main__":
    sys.exit(main())
